core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -612,16 +612,80 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
    * Provides the `ZManaged` effect with its required environment, which eliminates
    * its dependency on `R`.
    */
-  def provide(r: R)(implicit ev: NeedsEnv[R]): ZManaged[Any, E, A] =
+  def provide(r: R)(implicit ev: NeedsEnv[R]): Managed[E, A] =
     provideSome(_ => r)
 
+  /**
+   * Provides a layer to the `ZManaged`, which translates it to another level.
+   */
+  def provideLayer[E1 >: E, R0, R1 <: Has[_]](layer: ZLayer[R0, E1, R1])(implicit ev: R1 <:< R): ZManaged[R0, E1, A] =
+    layer.translate.value.flatMap(self.provide(_))
```

Comment:
How about `provideSomeManaged(layer.value.map(ev))`? Then we can revert the changes to `ZIO.scala` and `ZLayer.scala`.

Reply:
Yeah I do not mind removing the changes in `ZLayer.scala`, was trying to extract common behaviour but although I understood ZLayer that might been unnecessary. I still have the impression provideLayer for `ZIO` and `ZManaged` should look quite similar. Are you positive `provideSomeManaged(layer.value.map(ev))` suffices ?

Reply:
Looking at it @adamgfraser, your proposal looks quite clean. I was just wondering if this translation from one level to another could be moved to `ZLayer` but TBH I am fine with leaving that aside.

Reply:
Also, I am unsure whether your proposal would work. Should it not look like something like this, for us to have proper R translation? 
``` 
for {
      r0 <- ZManaged.environment[R0]
      r1 <- layer.value.provide(r0)
} yield ev(r1)).flatMap(self.provide(_)
```

Reply:
That isn't necessary because calling `value` on a `ZLayer[R0, E, R]` returns a `ZManaged[R0, E, R]` so it already expresses the ability to partially provide the environment. I checked that this compiles. We could add a test for it if you want.